aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -685,6 +685,513 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
   return result;
 }
 
+// helper methods for matrix_exp
+namespace {
+
+template <typename scalar_t, int ROW, int COL>
+using array2d = std::array<std::array<scalar_t, COL>, ROW>;
+
+// we consider 6 Taylor expansions of degree
+// 1, 2, 4, 8, 12, 18
+constexpr int total_n_degs = 6;
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+// Allocates a buffers of uninitialized or zero values
+// of shape [n_copies, a.size()]
+Tensor _allocate_buffer(const Tensor& a, int n_copies, bool is_zero = false) {
+  if (is_zero) {
+    return at::zeros(
+      {n_copies, a.size(0), a.size(1), a.size(2)},
+      a.options().memory_format(at::MemoryFormat::Contiguous)
+    );
+  }
+  else {
+    return at::empty(
+      {n_copies, a.size(0), a.size(1), a.size(2)},
+      a.options().memory_format(at::MemoryFormat::Contiguous)
+    );
+  }
+}
+
+// Makes `buffer` to store `num_matrices` number of matrices needed for
+// compute the matrix exponentials of different orders, i.e.
+// first `num_matrices` matrices from the list l := {I, A, A^2, A^3, A^6}
+// in a contiguous block of memory such that
+// buffer[0, ...] = l[0], // I
+// buffer[1, ...] = l[1], // A
+// ...
+// buffer[num_matrices - 1, ...] = l[num_matries - 1]
+void _fill_matrix_powers(Tensor& buffer, const Tensor& a, int num_matrices) {
+  // fill I
+  buffer.select(0, 0).copy_(
+    at::eye(a.size(-1), buffer.options()).expand_as(a)
+  );
```

Comment:
If there wasn't the `expand_as` call, you could just do this as `at::eye_out(buffer.select(0, 0), a.size(-1), buffer.options())` and skip the intermediate allocation. With the batch dimensions I'm less sure; this may be the best way to do it in this case.

Reply:
@zou3519 too bad we don't have batched eye!

Reply:
but we have diag_embed https://pytorch.org/docs/stable/generated/torch.diag_embed.html#torch.diag_embed

Reply:
Hmm,  but we don't have an out variant of diag embed... too bad.